File: web/src/lib/image/textCompositor.ts
import type { SKRSContext2D } from '@napi-rs/canvas';
import { RGBA, toCssColor } from './colorUtils';
import { BBox } from './regionReconstruction';

const MIN_FONT_PX = 8;
const FONT_STEP = 1;

// Languages conventionally written without spaces between words — wrap by
// character instead of by word for these targets.
const CHARACTER_WRAP_LANGS = new Set(['ja', 'ko', 'zh-CN', 'zh-TW']);

function wrapText(ctx: SKRSContext2D, text: string, maxWidth: number, byCharacter: boolean): string[] {
  const units = byCharacter ? Array.from(text) : text.split(/\s+/).filter(Boolean);
  const sep = byCharacter ? '' : ' ';
  const lines: string[] = [];
  let current = '';

  for (const unit of units) {
    const candidate = current ? current + sep + unit : unit;
    if (ctx.measureText(candidate).width <= maxWidth || current === '') {
      current = candidate;
    } else {
      lines.push(current);
      current = unit;
    }
  }
  if (current) lines.push(current);
  return lines;
}

function fitText(
  ctx: SKRSContext2D,
  text: string,
  box: BBox,
  targetLang: string,
  fontFamily: string,
  startFontPx: number,
): { fontPx: number; lines: string[] } {
  const maxWidth = box.x1 - box.x0;
  const maxHeight = box.y1 - box.y0;
  const byCharacter = CHARACTER_WRAP_LANGS.has(targetLang);

  let fontPx = Math.max(startFontPx, MIN_FONT_PX);
  for (; fontPx >= MIN_FONT_PX; fontPx -= FONT_STEP) {
    ctx.font = `${fontPx}px ${fontFamily}`;
    const lines = wrapText(ctx, text, maxWidth, byCharacter);
    const lineHeight = fontPx * 1.25;
    if (lines.length * lineHeight <= maxHeight || fontPx === MIN_FONT_PX) {
      return { fontPx, lines };
    }
  }
  ctx.font = `${MIN_FONT_PX}px ${fontFamily}`;
  return { fontPx: MIN_FONT_PX, lines: wrapText(ctx, text, maxWidth, byCharacter) };
}

/**
 * Draws `text` inside `box`, shrinking and wrapping it so it always stays
 * within the region originally occupied by the source text (requirement:
 * translated text must not spill outside the detected area).
 *
 * Translated text is always laid out horizontally, even when the *original*
 * text was vertical (e.g. Japanese manga dialogue): the target languages we
 * support (Portuguese, English, Spanish, ...) are not conventionally read
 * top-to-bottom, so horizontal wrapping inside the same box reads naturally
 * while still respecting the original placement/size of the bubble/caption.
 */
export function drawFittedText(
  ctx: SKRSContext2D,
  box: BBox,
  text: string,
  textColor: RGBA,
  targetLang: string,
  fontScale: number,
  fontFamily = 'sans-serif',
): void {
  if (!text.trim()) return;

  const naturalFontPx = Math.max(MIN_FONT_PX, Math.min(box.y1 - box.y0, 64));
  const desiredFontPx = Math.round(naturalFontPx * fontScale);
  const { fontPx, lines } = fitText(ctx, text, box, targetLang, fontFamily, desiredFontPx);

  ctx.font = `${fontPx}px ${fontFamily}`;
  ctx.textAlign = 'center';
  ctx.textBaseline = 'middle';

  const lineHeight = fontPx * 1.25;
  const blockHeight = lines.length * lineHeight;
  const startY = box.y0 + (box.y1 - box.y0 - blockHeight) / 2 + lineHeight / 2;
  const centerX = box.x0 + (box.x1 - box.x0) / 2;

  const strokeColor: RGBA = textColor[0] + textColor[1] + textColor[2] > 380 ? [0, 0, 0, 255] : [255, 255, 255, 255];
  ctx.lineWidth = Math.max(1, fontPx / 10);
  ctx.strokeStyle = `${toCssColor(strokeColor)}`;
  ctx.globalAlpha = 0.55;
  lines.forEach((line, i) => ctx.strokeText(line, centerX, startY + i * lineHeight));
  ctx.globalAlpha = 1;

  ctx.fillStyle = toCssColor(textColor);
  lines.forEach((line, i) => ctx.fillText(line, centerX, startY + i * lineHeight));
}

File: android/app/src/main/java/com/noctis/app/image/TextCompositor.kt
package com.noctis.app.image

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.text.Layout
import android.text.StaticLayout
import android.text.TextPaint
import com.noctis.app.model.BBox
import kotlin.math.min

private const val MIN_TEXT_SIZE_PX = 10f
private const val MAX_NATURAL_TEXT_SIZE_PX = 64f
private const val TEXT_SIZE_STEP = 1f

/**
 * Draws [text] inside [box], shrinking it (Android's [StaticLayout] already
 * wraps by word or by character as appropriate for the script) until it
 * fits vertically — the translated text must never spill outside the area
 * originally occupied by the source text.
 *
 * Like the web pipeline, translated text is always laid out horizontally
 * even if the original text was vertical (CJK): the target languages this
 * app supports are not conventionally read top-to-bottom.
 */
fun drawFittedText(canvas: Canvas, box: BBox, text: String, textColor: Int, fontScale: Float) {
    if (text.isBlank()) return
    val maxWidth = box.width.coerceAtLeast(1)
    val maxHeight = box.height.coerceAtLeast(1)

    val naturalSize = min(maxHeight.toFloat(), MAX_NATURAL_TEXT_SIZE_PX).coerceAtLeast(MIN_TEXT_SIZE_PX)
    var textSize = naturalSize * fontScale

    val paint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply { color = textColor }
    var layout = buildLayout(text, paint, textSize, maxWidth)

    while (layout.height > maxHeight && textSize > MIN_TEXT_SIZE_PX) {
        textSize = (textSize - TEXT_SIZE_STEP).coerceAtLeast(MIN_TEXT_SIZE_PX)
        layout = buildLayout(text, paint, textSize, maxWidth)
    }

    val strokeColor = if (Color.red(textColor) + Color.green(textColor) + Color.blue(textColor) > 380) Color.BLACK else Color.WHITE
    val strokePaint = TextPaint(paint).apply {
        style = Paint.Style.STROKE
        strokeWidth = (textSize / 10f).coerceAtLeast(1f)
        color = strokeColor
        alpha = 140
    }
    val strokeLayout = buildLayout(text, strokePaint, textSize, maxWidth)

    canvas.save()
    canvas.translate(box.x0.toFloat(), box.y0 + (maxHeight - layout.height) / 2f)
    strokeLayout.draw(canvas)
    layout.draw(canvas)
    canvas.restore()
}

private fun buildLayout(text: String, paint: TextPaint, textSize: Float, maxWidth: Int): StaticLayout {
    paint.textSize = textSize
    return StaticLayout.Builder.obtain(text, 0, text.length, paint, maxWidth)
        .setAlignment(Layout.Alignment.ALIGN_CENTER)
        .setLineSpacing(0f, 1.15f)
        .setIncludePad(false)
        .build()
}
